### PR TITLE
Fix an issue where configured backing services are not being deleted

### LINKED ISSUE #344: Regression: AppDeploymentDeleteServiceInstanceWorkflow leaks backend services on deletion
## Expected behavior

* As an app broker user provisioning brokered service without backing app (i.e. only backing service(s)), see #287
* in order to avoid backing service leak 
* I need backing services to be deleted when a brokered service is unprovisionned

## Observed behavior

Following #318, the AppDeploymentDeleteServiceInstanceWorkflow.deleteBackingServices() method is only attempting to delete backing services bound to applications, and hence skipping backing services when there is no associated backing services.

https://github.com/spring-cloud/spring-cloud-app-broker/blob/b0f2cbc5f4738cb0df8b5d06f3a9888c0188dc06/spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/workflow/instance/AppDeploymentDeleteServiceInstanceWorkflow.java#L75-L92

/CC @LittleBaiBai 

In addition, in case the provisioning completed with error such as the following, then this triggers a leak of backing services
   - the backing services failed to be bound to the backing applications
   - some backing services failed to be created
   - the backing application failed to starter

I understand that the following test enforces this behavior, i.e. that a non bound matching backing service instance does not get deleted upon brokered service deletion
https://github.com/spring-cloud/spring-cloud-app-broker/blob/b0f2cbc5f4738cb0df8b5d06f3a9888c0188dc06/spring-cloud-app-broker-integration-tests/src/test/java/org/springframework/cloud/appbroker/integration/DeleteInstanceWithServicesComponentTest.java#L105-L157

### PR DESCRIPTION
**Pull request by royclarkson**
When a service instance is deleted, all configured backing services will
be deleted as well as all services bound to any backing applications

Resolve #344

### CODE REVIEW

File: spring-cloud-app-broker-integration-tests/src/test/java/org/springframework/cloud/appbroker/integration/DeleteInstanceWithServicesComponentTest.java [hunk 1]
```diff
@@ -30,26 +30,22 @@
...
-	"spring.cloud.appbroker.services[0].apps[0].name=" + APP_NAME,
-	"spring.cloud.appbroker.services[0].apps[0].services[0].service-instance-name=" + BACKING_SI_NAME,
 	"spring.cloud.appbroker.services[0].services[0].service-instance-name=" + BACKING_SI_NAME,
 	"spring.cloud.appbroker.services[0].services[0].name=" + BACKING_SERVICE_NAME,
 	"spring.cloud.appbroker.services[0].services[0].plan=" + BACKING_PLAN_NAME
 })
 class DeleteInstanceWithServicesComponentTest extends WiremockComponentTest {
 
-	protected static final String APP_NAME = "app-delete-with-services";
+//	protected static final String APP_NAME = "app-delete-with-services";
```
Comment (LittleBaiBai) [inline spring-cloud-app-broker-integration-tests/src/test/java/org/springframework/cloud/appbroker/integration/DeleteInstanceWithServicesComponentTest.java:48]: Commented line to remove.

Comment (royclarkson) [reply spring-cloud-app-broker-integration-tests/src/test/java/org/springframework/cloud/appbroker/integration/DeleteInstanceWithServicesComponentTest.java:48]: Thanks!

File: spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/workflow/instance/AppDeploymentDeleteServiceInstanceWorkflow.java [hunk 2]
```diff
@@ -73,14 +75,8 @@ public AppDeploymentDeleteServiceInstanceWorkflow(BrokeredServices brokeredServi
...
-		return backingAppManagementService.getDeployedBackingApplications(request.getServiceInstanceId())
-			.flatMapMany(Flux::fromIterable)
-			.flatMap(backingApplication ->
-				Flux.fromIterable(backingApplication.getServices())
-					.map(servicesSpec -> BackingService.builder()
-						.serviceInstanceName(servicesSpec.getServiceInstanceName())
-						.build())
-					.collectList())
+		return collectConfiguredBackingServices(request)
+			.mergeWith(collectBoundBackingServices(request))
```
Comment (gberche-orange) [inline spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/workflow/instance/AppDeploymentDeleteServiceInstanceWorkflow.java:79]: Would there be duplicate BackingServices among those returned by `collectConfiguredBackingServices` and `collectBoundBackingServices` ?

Does this lead to polluting traces when trying to delete twice a service instance ?

Comment (royclarkson) [reply spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/workflow/instance/AppDeploymentDeleteServiceInstanceWorkflow.java:79]: Thanks for the feedback. Yes, it would result in multiple delete requests if a backing service instance is bound to an app. I'll consider how to eliminate those duplicates.

Comment (royclarkson) [reply spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/workflow/instance/AppDeploymentDeleteServiceInstanceWorkflow.java]: I've added a new commit that addresses the issue of duplicate service instances

File: spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/workflow/instance/AppDeploymentDeleteServiceInstanceWorkflow.java [hunk 3]
```diff
@@ -75,34 +77,52 @@ public AppDeploymentDeleteServiceInstanceWorkflow(BrokeredServices brokeredServi
...
+			.flatMap(backingApplication -> Mono.justOrEmpty(backingApplication.getServices())
+				.flatMapMany(Flux::fromIterable)
+				.flatMap(servicesSpec -> Mono.justOrEmpty(servicesSpec.getServiceInstanceName()))
+				.map(serviceInstanceName -> {
+					BackingService backingService = BackingService.builder()
+						.serviceInstanceName(serviceInstanceName)
+						.build();
+					if (!CollectionUtils.isEmpty(backingApplication.getProperties())) {
+						backingService.setProperties(Collections.singletonMap(DeploymentProperties.TARGET_PROPERTY_KEY,
+							backingApplication.getProperties().get(DeploymentProperties.TARGET_PROPERTY_KEY)));
```
Comment (LittleBaiBai) [inline spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/workflow/instance/AppDeploymentDeleteServiceInstanceWorkflow.java:122]: This line is pretty long and hard to follow. Would it be worth breaking the get target part out? Also, if the target property is null on backing application, we can probably skip setting it on backing service.

File: spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/workflow/instance/AppDeploymentDeleteServiceInstanceWorkflow.java [hunk 4]
```diff
@@ -75,34 +77,52 @@ public AppDeploymentDeleteServiceInstanceWorkflow(BrokeredServices brokeredServi
...
-					.serviceInstanceName(servicesSpec.getServiceInstanceName())
-					.build())
-				.collectList());
+			.flatMap(backingApplication -> Mono.justOrEmpty(backingApplication.getServices())
+				.flatMapMany(Flux::fromIterable)
+				.flatMap(servicesSpec -> Mono.justOrEmpty(servicesSpec.getServiceInstanceName()))
+				.map(serviceInstanceName -> {
+					BackingService backingService = BackingService.builder()
+						.serviceInstanceName(serviceInstanceName)
+						.build();
```
Comment (LittleBaiBai) [inline spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/workflow/instance/AppDeploymentDeleteServiceInstanceWorkflow.java:119]: Nit-picking: I would personally prefer moving the `.build()` to the return line and avoid using setter on backing service. It would give me a signal that this whole section is about building backing service. But the way it is right now is totally fine too.

File: spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/workflow/instance/AppDeploymentDeleteServiceInstanceWorkflow.java [hunk 5]
```diff
@@ -75,34 +77,52 @@ public AppDeploymentDeleteServiceInstanceWorkflow(BrokeredServices brokeredServi
...
-			.doOnEach(backingServices -> log.debug("Deleting backing services {} for {}/{}",
-				backingServices, request.getServiceDefinition().getName(), request.getPlan().getName()))
-			.flatMap(backingServicesProvisionService::deleteServiceInstance)
+		return collectBackingServices(request)
+			.collectList()
+			.flatMapMany(backingServices -> {
+				if (!CollectionUtils.isEmpty(backingServices)) {
+					return backingServicesProvisionService.deleteServiceInstance(backingServices);
+				}
+				return Flux.empty();
```
Comment (LittleBaiBai) [inline spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/workflow/instance/AppDeploymentDeleteServiceInstanceWorkflow.java:86]: `Flux.empty().collectList()` will result in an empty list instead of null, so I feel this null check here is not necessary.

File: spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/workflow/instance/AppDeploymentDeleteServiceInstanceWorkflow.java (hunk 3 above)
Comment (royclarkson) [reply spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/workflow/instance/AppDeploymentDeleteServiceInstanceWorkflow.java:122]: I'll rework it. Thanks!

File: spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/workflow/instance/AppDeploymentDeleteServiceInstanceWorkflow.java (hunk 5 above)
Comment (royclarkson) [reply spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/workflow/instance/AppDeploymentDeleteServiceInstanceWorkflow.java:86]: I got a NPE in the mapper without that check. I agree it's ugly and feels unnecessary. :-/

File: spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/workflow/instance/AppDeploymentDeleteServiceInstanceWorkflow.java (hunk 4 above)
Comment (royclarkson) [reply spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/workflow/instance/AppDeploymentDeleteServiceInstanceWorkflow.java:119]: makes sense. I'll move that logic to another function to make it read better.

Comment (royclarkson) [reply spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/workflow/instance/AppDeploymentDeleteServiceInstanceWorkflow.java]: reworked instead of moving to another function
